Upgrade to React-Native v0.61

hawkrives opened the pull request:
- [x] iOS: builds
- [x] Android: builds
- [x] Flow: passes

This was … really painless.

Changes:

- You need Xcode 11 to build from the command line (Xcode 10 can only build from the GUI)
- Android fonts are now copied at build time, instead of being part of our repo
- To fix type errors with React-Redux, I've replaced all `connect()` calls with the new `useDispatch` / `useSelector` hooks. Informal testing says that everything still works.
- To skip Cocoapods, you can `export FP_PODS=no` (for you, @rye)
- I added Cocoapods to our Gemfile, because it's a ruby dep
- We now use cocopods for iOS builds, because they're required
- We've switched from `react-native-sentry` to `@sentry/react-native`, because Sentry said to
- I removed `@frogpond/dimensions` because it just needed the "hasNotch" check from rn-device-info
- Added `useNativeDriver` where prompted
- assorted other changes for associated upgrades


Please test and lmk if things look broken! Otherwise we'll merge and test on nightly. It works at least enough for that.

hawkrives committed:
update modules to 0.61

hawkrives committed:
add new podspecs

hawkrives committed:
add cocoapods

hawkrives committed:
add automatic installation of cocoapods to postinstall

you can use `export FP_PODS=no` to skip cocoapods installation

hawkrives committed:
xcode did things

hawkrives committed:
update to @|sentry/react-native

hawkrives committed:
update to rn-gesture-handler@1.5.0

hawkrives committed:
update react-native-alternate-icons to support autolinking

hawkrives committed:
bump deployment target back down to 9.3

hawkrives committed:
downgrade rn-screens

hawkrives committed:
auto-install bundler packages after yarn finishes

hawkrives committed:
update pods repo before installing

hawkrives committed:
start updating android build

hawkrives committed:
run yarn install after installing bundler on macos

hawkrives committed:
rename sentry method to fix breadcrumbs

hawkrives committed:
install bundler@1.17.2

hawkrives committed:
update flow-bin

hawkrives committed:
update some flow things

hawkrives committed:
update all empty flow stubs

hawkrives committed:
add some more things

hawkrives committed:
update some of the simpler typedefs

drewvolz committed:
Clean up remaining linked libraries in the allaboutolaf target

hawkrives committed:
update redux calls

hawkrives committed:
update some more typedefs

hawkrives committed:
update flowconfig for proper version

hawkrives committed:
add .ignore file so ripgrep can ignore ios/Pods

hawkrives committed:
reformat MainApp.java

hawkrives committed:
remove uncalled method

hawkrives committed:
fix navigationOptions

hawkrives committed:
prevent installing sentry on debug

hawkrives committed:
fix warning from a useEffect call

hawkrives committed:
switch to react-native's SafeAreaView

hawkrives committed:
update Debug/List

hawkrives committed:
remove nested ScrollView / FlatLists from the search screen

hawkrives committed:
simplify finding the title in debug view

hawkrives committed:
remove unused props from CourseSearch/Search

hawkrives committed:
rename some internal methods in CourseSearch/Results

hawkrives committed:
break a thing into two lines

hawkrives committed:
tell rn-vector-icons to copy files, instead of copying them manually

hawkrives committed:
disable iOS 13 dark mode

hawkrives committed:
prettify

hawkrives committed:
begin updating to rn 0.61

rye commented on .github/workflows/analyze-and-build.yml:
```diff
@@ -62,8 +62,8 @@ jobs:
       - uses: actions/setup-node@master
         with: {node-version: ^12.0}
       - uses: actions/checkout@master
+      - run: gem install bundler:1.17.2 && bundle check || bundle install --frozen --path vendor/bundle
```
Why was this necessary? CocoaPods' version requirement on bundler of `~> 1.3`?

hawkrives replied:
If I don’t specify a version, then whatever `gem` wound up installing didn’t work, and the error message suggested this.

rye replied:
Ah okay, that clears things up.

rye commented on ios/Podfile:
```diff
@@ -0,0 +1,44 @@
+
+  use_native_modules!
+end
+
```
Our usual code style is tabs instead of spaces, if I remember right.

Also, please remove the trailing newline.

rye commented on scripts/pods.sh:
```diff
@@ -0,0 +1,16 @@
+
+if [[ $USE_PODS = 'yes' ]]; then
+	if [[ $(uname) = 'Darwin' ]]; then
+		bundle install --path .bundle
```
Is this required? I don't use `.bundle` in my Ruby projects, but if it's necessary in this case then I guess it's necessary.

hawkrives replied:
Where does bundler install packages by default? I don’t actually know; this is what my shell autocompletes when I do “bundle install” 😅

If I shouldn’t need this, then I’m happy to take it out.

rye replied:
You shouldn't need this; the default installation location is fine.

You can use `gem environment` to figure out where it's installing gems. `~/.gem` would be a place that it'd install.

rye commented on modules/timer/index.js:
```diff
@@ -80,6 +80,10 @@ export class Timer extends React.Component<Props, State> {
 		this.setState(() => ({loading: false}))
 	}
 
+	refresh = () => {
+		this._refresh()
+	}
+
```
This seems somewhat kludgey to me. Is there a particular reason why we need a shim?

hawkrives replied:
RN’s flow types require that functions passed to RefreshControl’s refresh prop be of the type `() => void`, and all Async functions return Promise<>, which causes type errors at each call site. 

My other alternative here is to typecast this.refresh as a variable in the render method here, but that requires that we cast through `any`, so I almost prefer this way. 

Idk.

hawkrives replied:
```js
refresh = () => {
	let start = Date.now()
	this.setState(() => ({loading: true}))

	this.updateTime()

	let tick = Promise.resolve(null)

	let elapsed = Date.now() - start
	if (elapsed < 500) {
		tick = tick.then(() => delay(500 - elapsed))
	}

	tick.then(() => {
		this.setState(() => ({loading: false}))
	})
}
```

That's my other option for this method lol

rye replied:
Alright, just curious.

rye reviewed (approved):
I gave this a pretty rapid once-over and it looks good enough to merge, just a couple of nitpicks and questions that don't really affect whether or not it can get shipped in nightly. (I'm perfectly happy to dogfood as long as it builds and makes it through detox on iOS.)

hawkrives committed:
reindent Podfile with tabs

hawkrives committed:
fix unused variable lint warnings

hawkrives committed:
remove trailing newline from Podfile

drewvolz reviewed (approved):
@rye's comments make sense. Thank you @hawkrives for putting in the legwork! 🥳

hawkrives committed:
attempt to switch to workspaces

hawkrives committed:
add "yarn fastlane" command

hawkrives committed:
remove auto-run of `pod repo update`

drewvolz committed:
reverting removal of appstore certificates in ios fastlane

rye committed:
pods.sh: Remove custom vendoring configuration

Just use the environment and trust rubygems to understand that.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Merge branch 'master' into rn0.61

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Merge branch 'master' into rn0.61

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

hawkrives committed:
rename GYM_PROJECT to XCODE_PROJECT

rye committed:
detox: Use the xcworkspace instead of xcodeproj

Similar to recent work on the fastlane configuration, this will instead
use the xcworkspace structure.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
app: Import react-native-gesture-handler before booting

It's hot garbage, but seems like it might be a solution to the issue,
still under investigation, at

    https://github.com/kmagiera/react-native-gesture-handler/issues/320

I do not know how this would fix things, but I am trying it.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
app: Add a TODO comment to remind us to stop using a weird fix

Per https://github.com/kmagiera/react-native-gesture-handler/issues/320

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>
